test/test_images/probe_helper/probe/helper.go

```diff
@@ -0,0 +1,402 @@
+		} else {
+			timeout = customTimeout
+		}
+	}
+	if timeout.Nanoseconds() > ph.MaxTimeoutDuration.Nanoseconds() {
+		logging.FromContext(ctx).Warnw("Desired timeout exceeds the maximum, clamping to maximum value", zap.Duration("timeout", timeout), zap.Duration("maximumTimeout", ph.MaxTimeoutDuration))
+		timeout = ph.MaxTimeoutDuration
+	}
+	return context.WithTimeout(ctx, timeout)
+}
+
+// withProbeEventLoggingContext attaches a logger to the context which contains
+// useful information about probe requests.
+func withProbeEventLoggingContext(ctx context.Context, event cloudevents.Event) context.Context {
+	logger := logging.FromContext(ctx)
+	logger = logger.With(zap.Any("event", map[string]interface{}{
+		"id":          event.ID(),
+		"source":      event.Source(),
+		"specversion": event.SpecVersion(),
+		"type":        event.Type(),
+		"extensions":  event.Extensions(),
+	}))
+	return logging.WithLogger(ctx, logger)
+}
+
+type cloudEventsFunc func(cloudevents.Event) cloudevents.Result
+
+// forwardFromProbe is the base forward probe request handler which is called
+// whenever the probe helper receives a CloudEvent through port PROBE_PORT or
+// through the specified probe port listener.
+func (ph *Helper) forwardFromProbe(ctx context.Context) cloudEventsFunc {
+	return func(event cloudevents.Event) cloudevents.Result {
+		// Attach important metadata about the event to the logging context.
+		ctx := withProbeEventLoggingContext(ctx, event)
+		logging.FromContext(ctx).Infow("Received probe request")
+
+		// Refresh the forward probe liveness time
+		ph.LivenessChecker.LastForwardEventTime.SetNow()
+
+		// Retrieve the probe handler based on the event type
+		pr, ok := ph.forwardProbeHandlers[event.Type()]
+		if !ok {
+			logging.FromContext(ctx).Warnw("Probe forwarding failed, unrecognized forward probe type")
+			return cloudevents.ResultNACK
+		}
+
+		// Ensure there is a targetpath CloudEvent extension
+		if _, ok := event.Extensions()[probeEventTargetPathExtension]; !ok {
+			logging.FromContext(ctx).Warnf("Probe forwarding failed, forward probe event missing '%s' extension", probeEventTargetPathExtension)
+			return cloudevents.ResultNACK
+		}
+
+		// Add timeout to the context
+		ctx, cancel := ph.withProbeTimeout(ctx, event)
+		defer cancel()
+
+		// Forward the probe event. This call is likely to be blocking.
+		if err := pr.Forward(ctx, event); err != nil {
+			logging.FromContext(ctx).Warnw("Probe forwarding failed", zap.Error(err))
+			return cloudevents.ResultNACK
+		}
+		return cloudevents.ResultACK
+	}
+}
+
+// receiveEvent is the base receiver probe request handler which is called
+// whenever the probe helper receives a CloudEvent through port RECEIVER_PORT or
+// through the specified receiver port listener.
+func (ph *Helper) receiveEvent(ctx context.Context) cloudEventsFunc {
```

Comment:
I think the receiver needs to verify that the path matches probeEventTargetPathExtension?

Reply:
One of the reasons why the clients are shared in this manner, or for the "if" checks, is that it allows us to easily override parts of the client initialization to use fake clients instead for testing purposes. The current setup is not critical, so we can definitely avoid tracking so much stuff in the helper struct, but we would still need to pass them to the helper initializer

Reply:
The initializer logic should live outside of helper. We can likely make this much cleaner with a dependency injection framework like wire. This can be a follow up.